graphql/tests/query.rs

```diff
@@ -2123,3 +2123,101 @@ fn deterministic_error() {
         assert_eq!(expected, serde_json::to_value(&result).unwrap());
     })
 }
+
+#[test]
+fn can_query_with_or_filter() {
+    const QUERY: &str = "
+    query {
+        musicians(where: { or: { name: \"John\", id: \"m2\" } }) {
+            name
+            id
+        }
+    }
+    ";
+
+    run_query(QUERY, |result, _| {
+        let exp = object! {
+            musicians: vec![
+                object! { name: "John", id: "m1" },
+                object! { name: "Lisa", id: "m2" },
+            ],
+        };
+        let data = extract_data!(result).unwrap();
+        assert_eq!(data, exp);
+    })
+}
+
+#[test]
+fn can_query_with_and_filter() {
+    const QUERY: &str = "
+    query {
+        musicians(where: { and: { name: \"John\", id: \"m2\" } }) {
+            name
+            id
+        }
+    }
+    ";
+
+    run_query(QUERY, |result, _| {
+        let exp = object! {
+            musicians: r::Value::List(vec![]),
+        };
+        let data = extract_data!(result).unwrap();
+        assert_eq!(data, exp);
+    })
+}
+
+#[test]
+fn can_query_with_or_and_filter() {
+    const QUERY: &str = "
+    query {
+        musicians(
+          where: {or:{and:{name: \"John\", id: \"m1\"}, mainBand: \"b2\" } }
+        ) {
+          name
+          id
+        }
+      }
+    ";
+
+    run_query(QUERY, |result, _| {
+        let exp = object! {
+            musicians: vec![
+                object! { name: "John", id: "m1" },
+                object! { name: "Tom", id: "m3" },
+            ],
+        };
+        let data = extract_data!(result).unwrap();
+        assert_eq!(data, exp);
+    })
+}
+
+#[test]
+fn can_query_with_and_and_filter() {
+    const QUERY: &str = "
+    query {
+        musicians(
+          where: {
+                or: {and: {name: \"John\", id: \"m1\" },
+                or: {and: {name: \"Lisa\", id: \"m2\" }
```

Comment:
I thought repeating the same key here was prohibited by the GraphQL spec, or am I misinterpreting this?

Also cc @azf20 for the syntax of writing `(name = 'John' and id = 'm1') or (name = 'Lisa' and id = 'm2')` As an alternative, we could maybe have `and` and `or` take either an object (`and: {name: "John", id: "m1"}`) or an array of objects (`or: [{and: {name: "John", id: "m1"}}, {and: {name: "Lisa", id: "m2"}}]`)

Reply:
It’s the way I formatted the query for test it may look like prohibited syntax but it isn’t :)
```graphql
query {
  musicians(
    where: {
      or: {
        and: { name: "John", id: "m1" }
        or: { and: { name: "Lisa", id: "m2" } }
      }
    }
  ) {
    id
    name
  }
}
```

Reply:
@saihaj I am a bit confused on what that query is doing, is it effectively:
```
(name == "John" AND id == "m1") OR (name == "Lisa" AND id == "m2")
```
Or is it something different? If so I think that syntax is a bit confusing

Reply:
yes @azf20 that is what the query is doing